### PR TITLE
SUSE Manager for Retail test suite

### PR DESCRIPTION
**Pull request by aaannz**
This PR contains SUSE Manager for Retail test suite and helpers.

Test suite is confined in `tests/suma` directory together with tests specific README.md and templates. Tests require at least 2 machines, designed for 4 (suma_master, suma_minion, suma_minion_terminal and suma_minion_terminal2) and tests SUSE Manager webui, salt integration and various salt deployments relevant to SUSE Manager for Retail.

Base test is located in `lib/sumatest.pm`.
Interesting helper is in `lib/selenium.pm` which contain helpers and configuration of selenium server to be installed on SUT so that selenium can be used directly from openQA test suite to test, in our case but not limited to, SUSE Manager WebUI.

### CODE REVIEW

**Comment by aaannz** [comment]
Test runs - http://sleposbuilder4.suse.cz/tests/overview?distri=suma_retail&version=12-SP2&build=0200&groupid=63

**Comment by asmorodskyi** [comment]
@foursixnine , @coolo - I not sure about suggested changes to cpan file . What do you think ?

**Comment by okurz** [comment]
are you sure this fits within os-autoinst-distri-opensuse and should not rather be in its own distribution and therefore its own repository?

**Comment by coolo** [comment]
I see no shared parts either - possibly we need to split distri-opensuse to be reusable for such distributions?

**Comment by okurz** [comment]
most likely. We can start here :)

**Comment by aaannz** [comment]
For start `lib/sumatests.pm` is derived from `x11test`. We use common functionality for building SUMA master and minion image to install SLES, then to manage booting the machines up to desktop. I don't think this deserves it's own distro. It's the similar approach as in slepos tests.

**Comment by okurz** [comment]
But what we mentioned would be to split out a helper library which could also include x11test

File: tests/suma/bind_formula.pm [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+        select_console 'x11', tags => 'suma_welcome_screen';
+
+        #    assert_and_click('suma-salt-menu');
+        #    assert_and_click('suma-salt-formulas');
+        #    assert_and_click('suma-branch-network-formula-details');
+        #    assert_screen('suma-branch-network-formula-details-screen');
+
+        # no form yet
+        #     assert_and_click('suma-systems-menu');
```
Comment (asmorodskyi) [inline tests/suma/bind_formula.pm]: please remove commented code from commit

**Comment by aaannz** [comment]
Yes, I get that. I have no problem with using different distri, although I don't get what is the reasoning for it. Note that I expect these tests to eventually be running on osd. This is a regular SUSE product in development.
Also tests do cover salt based dhcpd, pxe, dns, tftp configuration which are regular sles provided services. This isn't some product completely out of SLES scope.

Splitting openSUSE distri to different helper libraries sounds good, that way we would be able to publish standalone selenium helper library. But I'm not willing to do this split and infrastructure changes as part of this PR.

File: tests/suma/pxe_formula.pm [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    my ($self) = @_;
+    $self->register_barriers('pxe_formula', 'pxe_formula_finish');
+
+    my $srvdir = get_var('SERVER_DIR');
+
+    if (check_var('SUMA_SALT_MINION', 'branch')) {
+        $self->registered_barrier_wait('pxe_formula');
+        $self->registered_barrier_wait('pxe_formula_finish');
+    }
+    elsif (check_var('SUMA_SALT_MINION', 'terminal')) {
```
Comment (okurz) [inline tests/suma/pxe_formula.pm]: how about combining both if's? e.g. `if (get_var('SUMA_SALT_MINION') =~ /branch|terminal/)`

Comment (aaannz) [reply tests/suma/pxe_formula.pm]: I would rather keep this separated, mainly for convenience. Formula tests are scheduled on all SUTs and basically are following the same pattern - check for terminal, terminal2, branch or else it is master. 'minion' and 'branch' are different machines with different roles so if any test is added to either one of them, this tests will be definitely different for terminal and branch minions.

**Comment by aaannz** [comment]
I'm sorry for taking it so long to respond. I was busy with other tasks. Anyway, here goes updated version with new tests, removed commented code and some FIXMEs and TODOs. Some I left, because the product does not support yet all the functionality.

Latest run - http://sleposbuilder4.suse.cz/tests/overview?distri=suma_retail&version=12-SP2&build=manual&groupid=63 there seems to be some font change in the terminal tests, I'll need to investigate that.

File: cpanfile [hunk 3]
```diff
@@ -7,6 +7,9 @@ requires 'XML::Simple';
 requires 'IO::File';
 requires 'List::Util';
 requires 'LWP::Simple';
+requires 'Selenium::Remote::Driver';
+requires 'Selenium::Chrome';
+requires 'Selenium::Waiter';
```
Comment (okurz) [inline cpanfile]: Hm, I'm not so sure about this added dependency. Is it too heavy weight in some situations or ok in general?

File: tests/suma/shutdown.pm [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    send_key('ctrl-alt-f10');
+    send_key('esc');
+    if (check_var('VIRSH_VMM_FAMILY', 'xen')) {
+        assert_shutdown_and_restore_system($action);
+    }
+    else {
+        assert_shutdown(480) if $action eq 'poweroff';
+        reset_consoles;
+    }
+}
```
Comment (okurz) [inline tests/suma/shutdown.pm]: There is a helper function power_action in lib/utils. I think that one should be used instead. Extend that one of necessary

**Comment by foursixnine** [comment]
@asmorodskyi  They should not be a major problem. however, these are specific for testing with selenium, and while it doesn' t hurt... 

Does it really make sense to have tests for web applications within the os-autoinst-distri-opensuse?, having a split could be the kickstart to have other teams to also test web apps within openQA and benefit from the availability of different architectures and focus on their application testing rather than problems with other parts of a normal openQA test.  But that' s another story

But as for these changes I agree with what has been discussed before in this regard: Changes are already there, but this PR is already too big to add even more complexity... and as a proof of concept might be ok I guess. 

But I would be a bit careful and come up with a convention to know what is using selenium. (But that's me, being paranoic)

**Comment by ldevulder** [comment]
The Selenium part could be useful for HA QA also for testing the Hawk web interface (cluster and resource configuration).

**Comment by ldevulder** [comment]
Is there any news on this PR?

As I said, it could be useful for some HA tests as well (Hawk Web interface), but I agree with @okurz and @foursixnine that it's a huge PR and may add complexity. But on the other hand, it could be good to use openQA for all our tests, and use Selenium inside openQA to test Hawk. But of course, it's possible to use Selenium apart.

To sum up: I would like to know if this PR has any chance to be merged, or if I need to use Selenium apart.

Thanks :)

**Comment by aaannz** [comment]
I'll split selenium part into separate PR, there are numerous other changes in this test suite and I didn't get to rebase it during holidays as I planned.

**Comment by asmorodskyi** [comment]
@aaannz  probably you can close this PR , since it will never be merged anyway

**Comment by aaannz** [comment]
True, many things changed.

**Comment by okurz** [comment]
I wouldn't be so sure about it. We merged already some part of it in #4412 so now it is easier to see the change if you rebase but feel free to open a new PR with the left changes

**Comment by asmorodskyi** [comment]
actually I agree with @okurz here . I think main reasons why this PR was not merged you already merge in another PR. So what is left here most probably would be easily merged . But also I think this should be done in new PR because this one contains tons of discussion not related to what left to merge

**Comment by aaannz** [comment]
Unfortunately I don't have time for upstreaming effort right now. Even the simple #4412 took way more time than should. Then we are still doing substantial changes in test suite all the time. So maybe after release after summer things will stabilize and I'll do new PR. At least as a demo how selenium tests can be used.

**Pull request closed without merging**